### PR TITLE
Localization and internationalization in android

### LINKED ISSUE #5011: [Android]Localization and internationalization in android
[Android]Localization and internationalization in android

#### introduction
Android is used on a wide range of devices in a variety of countries. Your app should handle text, audio files, numbers, money, and visuals in ways that are acceptable to the locations where your app is utilized to reach the most customers.

This article will go through the Android localization process in great depth, as well as how to apply best practices to your translation efforts.

#### Key takeaways 
The key takeaways are:-
- localizing strings,images and simple media
- understanding localization resources in android
- organizing resources with configuration types and qualifiers
- providing localized resources considering various factors
- understanding how SDK deals with your localized resources
- Applying best practices to your localization app effort

#### Article quality
This article will aim to guide readers on the best practices to apply on a localized app and show the relevant and best way to perform localization in android application and resources.

#### References
N/A

### PR DESCRIPTION
**Pull request by nia-vee**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #5011

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Review by BrianaNzivu — CHANGES_REQUESTED** [review]
Hey @nia-vee  👋 
Welcome to Section - EngEd. 🎉 
We are happy to have you here.
Just a few things I would love to point out before we proceed with the preliminary revisions:
- I am noticing that your images are not 600 x 400 px. Please scale down your images. You can use [BeFunky](https://www.befunky.com/) to do this,
- Your article has quite a lot of mistakes. I suggest you use [Grammarly](https://app.grammarly.com/) to make these corrections. 
- Please structure your article as required by Section. You can check out a tutorial's structures [here](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md).

Ping me when done.

Kind Regards.

**Comment by nia-vee** [comment]
Hello @BrianaNzivu i have made the requested changes 
Thanks

**Comment by BrianaNzivu** [comment]
Hey @nia-vee 👋 .
This article still has a lot of grammatical errors. Please cross-read it and use [Grammarly](https://app.grammarly.com/) to go through it to avoid closure of the closure of this PR.

**Please note: Our Peer Reviewer and Content Moderator teams do NOT provide any revisions services. All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner.**

**Comment by BrianaNzivu** [comment]
Hey @nia-vee 👋🏾.
Everything OK?
Need some help?

**Comment by nia-vee** [comment]
Hello @BrianaNzivu Have sent a message on slack.
Thanks in advance and am looking forward to get some help..

**Comment by BrianaNzivu** [comment]
Tech review done.

**Comment by WanjaMIKE** [comment]
Plagiarism check done

**Review by BrianaNzivu — APPROVED** [review]
Hey @nia-vee 👋 .

Very good job on the article. It is a very interesting read and has minimal mistakes. This is very impressive considering you are a first-time contributor 👏 .

Just a few things I'd like to point out:
- I suggest that you check out our resources page which has links to various resourceful tools for writing just to minimize the changes being made by the reviewer.
- Lastly please go through ALL the corrections I have made so that you can not repeat the same mistakes.

That's it! :smile: 

A final review will be done now by a content moderator.
Otherwise, congratulations on publishing your first article🍾.

In case of any question or clarification (currently or in the future), do not hesitate to reach out via [Slack](https://app.slack.com/client/T5X14SH7G/D01A6MT673R) or my personal email (briananzivu@gmail.com). 
I am **ALWAYS** here to help.

It has been a pleasure reading and reviewing this article.

Looking forward to seeing you contribute more to EngEd. 🚀

Once again good job! 👍 

Kind Regards.

*P.s Wishing you a happy new year and a prosperous 2022.*

File: content/articles/localization-and-internationalization-in-android/index.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/localization-and-internationalization-in-android/hero.jpg
+    alt: Localization and internationalization in Android Hero Image
+---
+### Introduction
+Localization is the process of modeling applications to a specific locale, market, or device. Localization is said to be translation.
+<!--more-->
+Globalization is the process of building and developing a software product that functions in multiple cultures and locales.
```
Comment (WanjaMIKE) [inline content/articles/localization-and-internationalization-in-android/index.md:20]: Globalization has a wider meaning. Could you rephrase

**Comment by nia-vee** [comment]
Hello @WanjaMIKE have given the clear definition 
Thank you.

File: content/articles/localization-and-internationalization-in-android/index.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+- In cases where you want to use the same resource but with another name within your application. Avoid duplication of resources but rather use aliases instead, which reduces the growth of APK size.
+- When doing translation of strings, use professional translation services. To make sure the correct meaning is conveyed when each string is displayed.
+
+### Precautions when localizing your application
+- Avoid translation engines that could give you the wrong output of the string you want to localize. However, instead, use professional translation services.
+- Do not over-localize your application. Ensure you select the resources you want to localize according to the configuration type and not all the available resources
+- Think of localization in the early stage of application development.
+
+### Conclusion
+In this tutorial, we have learned localization, globalization, and internationalization. How the Android SDK deals with localized resources. We also learned the localization of strings and images. Best practices to follow when localization and precautions are taken when localizing your application.
```
Comment (WanjaMIKE) [inline content/articles/localization-and-internationalization-in-android/index.md:189]: `How the Android SDK deals with localized resources` this sentence is oddly placed. Rephrase

File: content/articles/localization-and-internationalization-in-android/index.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+### Localizing images 
+Its vital for an image to display the right content according to the context it is being utilized. For this reason, localization of images should be considered.
+
+This feature can be implemented as follows:
+
+#### Step 1 - Creating a new image subdirectory
+`Right-click` on the `resource` folder and navigate to the new `select android resource` directory.
+
+#### Step 2 - Naming newly created image subdirectory
+Name the directory similar to the available image directory and select drawable on the resource type on the wizard that appears under the available qualifier.
```
Comment (WanjaMIKE) [inline content/articles/localization-and-internationalization-in-android/index.md:146]: This section seems rushed. Highlight things such as folder names etc. Explain it more clearly.

File: content/articles/localization-and-internationalization-in-android/index.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+This feature can be implemented as follows:
+
+#### Step 1 - Creating a new image subdirectory
+`Right-click` on the `resource` folder and navigate to the new `select android resource` directory.
+
+#### Step 2 - Naming newly created image subdirectory
+Name the directory similar to the available image directory and select drawable on the resource type on the wizard that appears under the available qualifier. 
+
+Select locale hit the forward arrows to choose the appropriate qualifier, and hit ok. You will notice that the directory name is appended with the specified qualifier symbol, as shown below.
```
Comment (WanjaMIKE) [inline content/articles/localization-and-internationalization-in-android/index.md:148]: Replace hit with another synonym. Minimize repetition.

File: content/articles/localization-and-internationalization-in-android/index.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+![create_drawable_file](/engineering-education/localization-and-internationalization-in-android/create_drawable_file.png)
+
+#### Step 3 - Opening the created drawable subdirectory in your computer file
+Right-click on the resource folder and open it in your files
+
+#### Step 4: Locate the drawable subdirectory in your files and add the required image
+Locate the newly created drawable subdirectory and add an image you wish to display for the specified locale. Ensure the image added within the new drawable folder has a similar name as that within the original drawable folder.
```
Comment (WanjaMIKE) [inline content/articles/localization-and-internationalization-in-android/index.md:156]: Highlight key points like `this`

File: content/articles/localization-and-internationalization-in-android/index.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+When you run your application, expect the below output.
+
+![english](/engineering-education/localization-and-internationalization-in-android/english.png)
+
+![swahili](/engineering-education/localization-and-internationalization-in-android/swahili.png)
+
+
+### Media files
+There are always two types of media files in the Android application:
+-Text-to-Speech is audio that speaks out any text within the `string.xml` or hardcoded in the application.
```
Comment (WanjaMIKE) [inline content/articles/localization-and-internationalization-in-android/index.md:170]: is text-to-speech an audio or engine?

File: content/articles/localization-and-internationalization-in-android/index.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+-Text-to-Speech is audio that speaks out any text within the `string.xml` or hardcoded in the application. 
+
+It uses text-to-speech engine, which speaks out any text provided, which handles the localization process effectively as it speaks out its text.
+
+- Recorded Audio files - These are any audio file example, MP3, which are embedded in your application and played at different times.
+
+### Best Practices for localization
+- Provide a complete set of all the default resources to localize the application.
+- Do not think of the localization process at the final application development stage. Instead, think of the localization process at the design stage of the application, which makes it easy to localize.
+- Ensure that the qualifier naming rule is followed. The best way to follow this naming rule is by creating an alternate resource directory using the new resource directory wizard and not doing it manually.
```
Comment (WanjaMIKE) [inline content/articles/localization-and-internationalization-in-android/index.md:179]: Use another synonym for `follow`. Avoid repetition

File: content/articles/localization-and-internationalization-in-android/index.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+- In cases where you want to use the same resource but with another name within your application. Avoid duplication of resources but rather use aliases instead, which reduces the growth of APK size.
+- When doing translation of strings, use professional translation services. To make sure the correct meaning is conveyed when each string is displayed.
+
+### Precautions when localizing your application
```
Comment (WanjaMIKE) [inline content/articles/localization-and-internationalization-in-android/index.md:183]: Is this section similar to `Best Practices ...`?. I think so.

File: content/articles/localization-and-internationalization-in-android/index.md [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+![swahili](/engineering-education/localization-and-internationalization-in-android/swahili.png)
+
+
+### Media files
+There are always two types of media files in the Android application:
+-Text-to-Speech is audio that speaks out any text within the `string.xml` or hardcoded in the application. 
+
+It uses text-to-speech engine, which speaks out any text provided, which handles the localization process effectively as it speaks out its text.
+
+- Recorded Audio files - These are any audio file example, MP3, which are embedded in your application and played at different times.
```
Comment (WanjaMIKE) [inline content/articles/localization-and-internationalization-in-android/index.md:174]: No need to define recorded audio files.

File: content/articles/localization-and-internationalization-in-android/index.md [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+- In cases where you want to use the same resource but with another name within your application. Avoid duplication of resources but rather use aliases instead, which reduces the growth of APK size.
+- When doing translation of strings, use professional translation services. To make sure the correct meaning is conveyed when each string is displayed.
+- Avoid translation engines that could give you the wrong output of the string you want to localize. However, instead, use professional translation services.
+- Do not over-localize your application. Ensure you select the resources you want to localize according to the configuration type and not all the available resources
+- Think of localization in the early stage of application development.
+
+### Conclusion
+In this tutorial, we have learned localization, globalization, and internationalization. How Android SDK deals with localized resources. We also learned the localization of strings and images. Best practices to follow when localizing your application.
+
+To get the full implementation of the project, visit this [GitHub Repository](https://github.com/nia-vee/LocalizationInternationalizationDemo)
```
Comment (WanjaMIKE) [inline content/articles/localization-and-internationalization-in-android/index.md:187]: The link you included here does not work.

**Comment by nia-vee** [comment]
Hello? @WanjaMIKE 
Kindly don't remove the link i have fixed it i think i had included the URl wrongly but i have check on it is now working.
Thank you

**Comment by WanjaMIKE** [comment]
@nia-vee Kindly, take note of the edits to improve in your future submission.

**Comment by WanjaMIKE** [comment]
Merged #5554 into master
